core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -198,14 +211,19 @@ class TransactionCoordinator(brokerId: Int,
           Left(Errors.CONCURRENT_TRANSACTIONS)
 
         case CompleteAbort | CompleteCommit | Empty =>
-          val transitMetadata = if (txnMetadata.isProducerEpochExhausted) {
+          val transitMetadataResult = if (txnMetadata.isProducerEpochExhausted) {
             val newProducerId = producerIdManager.generateProducerId()
-            txnMetadata.prepareProducerIdRotation(newProducerId, transactionTimeoutMs, time.milliseconds())
+            Right(txnMetadata.prepareProducerIdRotation(newProducerId, transactionTimeoutMs, time.milliseconds()))
```

Comment:
I think this case should only be allowed if `currentProducerEpoch` is equal to `Short.MaxValue`?